### PR TITLE
feat: fixing seek to node by pattern

### PR DESCRIPTION
**Pull request by OperationValkyrie**
Fixed the SeekToNodeByPattern by moving to template functions. Should be good to replace other SeekToNode functions.

### CODE REVIEW

File: include/ParserCommon.h [hunk 1]
```diff
@@ -169,7 +172,109 @@ bool NodeHasAttributeContent(xmlNodePtr root_node, const std::string &target_con
 
 ParserXmlNodeSeek SeekToNodeByName(xmlNodePtr root_node, const std::string &name);
 ParserXmlNodeSeek SeekToNodeByNameRecursive(xmlNodePtr root_node, const std::string &name);
-ParserXmlNodeSeek SeekToNodeByPattern(xmlNodePtr root_node, int num, ...);
+
+std::string TrimWhitespace(const std::string &target_string);
+std::string getXmlNodePrint(xmlNodePtr root_node);
+void getXmlNodePrintHelper(xmlNodePtr root_node, std::stringstream& ss, int indent);
```
Comment (benjaminchang23) [inline include/ParserCommon.h]: Captial first letter GetXmlNodePrintHelper

File: src/ParserRR.cc [hunk 2]
```diff
@@ -149,6 +149,15 @@ void ParserRR::FindChapterNodes(xmlNodePtr root_node)
 
 void ParserRR::FindMetaData(xmlNodePtr root_node)
 {
+    ParserXmlNodeSeek head_seek = SeekToNodeByName(root_node, "head");
+
+    if (!head_seek.found)
+    {
+        std::cout << "Could not find head, exiting" << std::endl;
```
Comment (benjaminchang23) [inline src/ParserRR.cc]: pretty sure this doesn't actually exit

File: src/ParserCommon.cc [hunk 3]
```diff
@@ -374,91 +374,58 @@ ParserXmlNodeSeek SeekToNodeByNameRecursive(xmlNodePtr root_node, const std::str
...
+std::string TrimWhitespace(const std::string& target_string)
+{
+    auto leading_pos = target_string.find_first_not_of(" \t\r\n\0");
+    auto trailing_pos = target_string.find_last_not_of(" \t\r\n\0");
+    if (leading_pos == std::string::npos && trailing_pos == std::string::npos)
     {
-        result.seek_node = root_node;
-        result.found = true;
-        return result;
+        return "";
```
Comment (benjaminchang23) [inline src/ParserCommon.cc]: this got moved to the black-library-common library

File: src/ParserCommon.cc [hunk 4]
```diff
@@ -374,91 +374,58 @@ ParserXmlNodeSeek SeekToNodeByNameRecursive(xmlNodePtr root_node, const std::str
     return seek;
 }
 
-ParserXmlNodeSeek SeekToNodeByPattern(xmlNodePtr root_node, int num, ...)
+bool SeekToNodeByPatternHelper(xmlNodePtr root_node)
```
Comment (benjaminchang23) [inline src/ParserCommon.cc:377]: this always returns true?

Comment (OperationValkyrie) [reply src/ParserCommon.cc:377]: Its the end case of the template recursive calls. So yes. A SeektoNodeByPatternHelper without any patterns should return true.
Calls goes like
SeekToNodeByPatternHelper(node, PATTERN1, match1, PATTERN2, match2) - checks pattern1, if passes calls next
SeekToNodeByPatternHelper(node, PATTERN2, match2) - checks pattern 2, if passes calls next
SeekToNodeByPatternHelper(node) - always returns true.

File: include/ParserCommon.h [hunk 5]
```diff
@@ -169,7 +172,105 @@ bool NodeHasAttributeContent(xmlNodePtr root_node, const std::string &target_con
 
 ParserXmlNodeSeek SeekToNodeByName(xmlNodePtr root_node, const std::string &name);
 ParserXmlNodeSeek SeekToNodeByNameRecursive(xmlNodePtr root_node, const std::string &name);
-ParserXmlNodeSeek SeekToNodeByPattern(xmlNodePtr root_node, int num, ...);
+
+template<typename ... Args>
+ParserXmlNodeSeek SeekToNodeByPattern(xmlNodePtr root_node, Args... args)
+{
+    ParserXmlNodeSeek result;
+    if(!root_node) {
```
Comment (benjaminchang23) [inline include/ParserCommon.h]: fix the spacing here

**Review by benjaminchang23 — APPROVED** [review]

**Pull request merged**